autoendpoint/src/routers/fcm/client.rs

```diff
@@ -18,42 +21,155 @@ pub struct FcmClient {
     endpoint: Url,
     timeout: Duration,
     max_data: usize,
-    auth: DefaultAuthenticator,
-    http: reqwest::Client,
+    fcm_authenticator: Option<DefaultAuthenticator>,
+    http_client: reqwest::Client,
+    server_credential: FcmServerCredential, // Only used for legacy GCM
 }
 
 impl FcmClient {
     /// Create an `FcmClient` using the provided credential
     pub async fn new(
         settings: &FcmSettings,
-        credential: FcmCredential,
+        server_credential: FcmServerCredential,
         http: reqwest::Client,
     ) -> std::io::Result<Self> {
-        let key_data = serde_json::from_str::<ServiceAccountKey>(&credential.credential)?;
-        let auth = ServiceAccountAuthenticator::builder(key_data)
-            .build()
-            .await?;
-
+        // `map`ping off of `serde_json::from_str` gets hairy and weird, requiring
+        // async blocks and a number of other specialty items. Doing a very stupid
+        // json detection does not. GCM keys are base64 values, so  `{` will never
+        // appear in a GCM key. FCM keys are serialized JSON constructs.
+        // These are both set in the settings and come from the `credentials` value.
+        let auth = if server_credential.server_access_token.contains('{') {
+            trace!(
+                "Reading credential for {} from string...",
+                &server_credential.project_id
+            );
+            let key_data =
+                serde_json::from_str::<ServiceAccountKey>(&server_credential.server_access_token)?;
+            Some(
+                ServiceAccountAuthenticator::builder(key_data)
+                    .build()
+                    .await?,
+            )
+        } else {
+            // check to see if this is a path to a file, and read in the credentials.
+            if Path::new(&server_credential.server_access_token).exists() {
+                warn!(
+                    "Reading credential for {} from file...",
+                    &server_credential.project_id
+                );
+                let mut file = File::open(&server_credential.server_access_token)?;
+                let mut content = String::new();
+                file.read_to_string(&mut content)?;
```

Comment:
Rust added this helper at some point
```suggestion
                let content = fs::read_to_string(&server_credential.server_access_token)?;
```